blockindex/sgd.go

```diff
@@ -0,0 +1,235 @@
+	if kv == nil {
+		panic("nil kvstore")
+	}
+	if percentage > 100 {
+		panic("percentage should be less than 100")
+	}
+	return &sgdRegistry{
+		kvStore:    kv,
+		percentage: percentage,
+		kvCache:    kvCache,
+	}
+}
+
+// Start starts the SGDIndexer
+func (sgd *sgdRegistry) Start(ctx context.Context) error {
+	return sgd.kvStore.Start(ctx)
+}
+
+// Stop stops the SGDIndexer
+func (sgd *sgdRegistry) Stop(ctx context.Context) error {
+	return sgd.kvStore.Stop(ctx)
+}
+
+// Height returns the current height of the SGDIndexer
+func (sgd *sgdRegistry) Height() (uint64, error) {
+	h, err := sgd.kvStore.Get(_sgdToHeightNS, _sgdCurrentHeight)
+	if err != nil {
+		// if db not exist, return 0, nil, after PutBlock, the height will be increased
+		if errors.Is(err, db.ErrNotExist) {
+			return 0, nil
+		}
+		return 0, err
+	}
+	return byteutil.BytesToUint64BigEndian(h), nil
+}
+
+// PutBlock puts a block into SGDIndexer
+func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
+	var (
+		index    *indexpb.SGDIndex
+		r        *action.Receipt
+		actType  sgdActType
+		contract string
+		ok       bool
+	)
+	if blk == nil {
+		return errors.New("empty block")
+	}
+	b := batch.NewBatch()
+	receipts := getReceiptsFromBlock(blk)
+	for _, selp := range blk.Actions {
+		act := selp.Action()
+		actHash, err := selp.Hash()
+		if err != nil {
+			continue
+		}
+		switch act := act.(type) {
+		case *action.Execution:
+			r, ok = receipts[actHash]
+			if !ok || r.Status != uint64(iotextypes.ReceiptStatus_Success) {
+				continue
+			}
+			if r.ContractAddress != "" {
+				actType = _sgdActCreate
+				contract = r.ContractAddress
+			} else {
+				actType = _sgdActCall
+				contract = act.Destination()
+			}
+			sender, _ := address.FromBytes(selp.SrcPubkey().Hash())
+			sgdAct := sgdAct{
+				actType:    actType,
+				sender:     sender,
+				contract:   contract,
+				createTime: blk.Header.Timestamp(),
+			}
+			if index, err = sgd.actToIndex(sgdAct); err != nil {
+				return err
+			}
+			if err := sgd.putIndex(b, index); err != nil {
+				return err
+			}
+		default:
+		}
+	}
+	b.Put(_sgdToHeightNS, _sgdCurrentHeight, byteutil.Uint64ToBytesBigEndian(blk.Height()), "failed to put current height")
+	return sgd.kvStore.WriteBatch(b)
+}
+
+func (sgd *sgdRegistry) actToIndex(sgdAct sgdAct) (*indexpb.SGDIndex, error) {
+	var (
+		sgdIndex *indexpb.SGDIndex
+		err      error
+	)
+	switch sgdAct.actType {
+	case _sgdActCreate:
+		sgdIndex = &indexpb.SGDIndex{
+			Contract:   sgdAct.contract,
+			Deployer:   sgdAct.sender.String(),
+			CreateTime: sgdAct.createTime.Unix(),
+		}
+	case _sgdActCall:
+		sgdIndex, err = sgd.GetSGDIndex(sgdAct.contract)
+		if err != nil {
+			//make sure running when the contract is not exist
+			sgdIndex = &indexpb.SGDIndex{
+				Contract: sgdAct.contract,
+			}
+		}
+		sgdIndex.CallTimes++
+	}
+	return sgdIndex, nil
+
+}
+
+func (sgd *sgdRegistry) putIndex(b batch.KVStoreBatch, sgdIndex *indexpb.SGDIndex) error {
```

Comment:
should take `sgdAct` as input type, do the conversion to `indexpb.SGDIndex` inside

Reply:
removed `sgdAct`